File: Python_Ruby/class_member/4.rb
class Calculator
    @@_history = []

    def Calculator.history()
        for item in @@_history
            p item
        end
    end

    def initialize(v1, v2)
        @v1 = v1
        @v2 = v2
    end

    def setV1(v)
        if v.is_a?(Integer)
            @v1 = v
        else
            p 'you tried changing v1 by not integer'
        end
    end

    def add()
        result = @v1 + @v2
        @@_history.push("add : #{@v1}+#{@v2}=#{result}")
        return result
    end

    def subtract()
        result = @v1 - @v2
        @@_history.push("subtract : #{@v1}-#{@v2}=#{result}")
        return result
    end
end

class MultipleCalculator < Calculator
    def multiply()
        result = @v1 * @v2
        @@_history.push("multiply : #{@v1}*#{@v2}=#{result}")
        return result
    end
end

class DivideCalculator < MultipleCalculator
    def divide()
        result = @v1 / @v2
        @@_history.push("divide : #{@v1}/#{@v2}=#{result}")
        return result
    end
end

c1 = MultipleCalculator.new(10, 10)
p 'add', c1.add()
p 'subtract', c1.subtract()
p 'multiply', c1.multiply()

c2 = DivideCalculator.new(20, 10)
p 'add', c2.add()
p 'subtract', c2.subtract()
p 'multiply', c2.multiply()
p 'divide', c2.divide()

Calculator.history()

File: Python_Ruby/class_member/4.py
class Calculator(object):
    _history = []

    @classmethod
    def history(cls):
        print("---History of the Calculator---")
        for item in Calculator._history:
            print(item)

    def __init__(self, v1, v2):
        self.v1 = v1
        self.v2 = v2

    def setV1(self, v):
        if(isinstance(v, int)):
            self.v1 = v
        else:
            print('You tried changing v1 by not integer')

    def getV1(self):
        return self.v1

    def add(self):
        result = self.v1 + self.v2
        Calculator._history.append("add : %d+%d=%d" % (self.v1, self.v2, result))
        return result
    def subtract(self):
        result = self.v1 - self.v2
        Calculator._history.append("subtract : %d-%d=%d" % (self.v1, self.v2, result))
        return result

class MultipleCalculator(Calculator):
    def multiply(self):
        result = self.v1 * self.v2
        Calculator._history.append("multiply : %d*%d=%d" % (self.v1, self.v2, result))
        return result

class DivideCalculator(MultipleCalculator):
    def divide(self):
        result = self.v1 / self.v2
        Calculator._history.append("divide : %d/%d=%d" % (self.v1, self.v2, result))
        return result

c1 = MultipleCalculator(10, 10)
print('add', c1.add())
print('subtract', c1.subtract())
print('multiply', c1.multiply())

c2 = DivideCalculator(20, 10)
print('add', c2.add())
print('subtract', c2.subtract())
print('multiply', c2.multiply())
print('divide', c2.divide())

Calculator.history()